crates/kelk/src/storage/bst/mod.rs

```diff
@@ -88,70 +90,93 @@ where
-            if node.key.gt(key) {
-                is_right = true;
-                offset = node.right;
-            } else if node.key.lt(key) {
-                is_right = false;
-                offset = node.left;
-            }
-
-            if offset.eq(&0) {
-                return Ok(None);
-            }
-
-            node = self.storage.read(offset)?;
-        }
-        let result = node.value;
+        while offset != 0 {
+            let mut node: Node<K, V> = self.storage.read(offset)?;
 
-        // Case 1: If node to be deleted has only one child
-        if node.left.eq(&0) {
-            offset = node.right;
-        } else if node.right.eq(&0) {
-            offset = node.left;
-        } else {
-            // Case 2: If node to be deleted has two children
-            // Get the inorder successor (smallest in the right subtree)
-            let mut min_val = node.right;
-            let mut min_node: Node<K, V> = self.storage.read(min_val)?;
-            while !min_node.left.eq(&0) {
-                min_val = min_node.left;
-                min_node = self.storage.read(min_val)?;
-            }
+            match key.cmp(&node.key) {
+                Ordering::Less => {
+                    node_parent_info = Some((offset, node.clone()));
+                    offset = node.left;
+                    is_right = false;
+                }
+                Ordering::Greater => {
+                    node_parent_info = Some((offset, node.clone()));
+                    offset = node.right;
+                    is_right = true;
+                }
+                Ordering::Equal => {
+                    let result = node.value.clone();
+                    match (node.left, node.right) {
+                        // The node has no children or only one child
+                        (0, 0) | (_, 0) | (0, _) => {
+                            match &mut node_parent_info {
+                                Some((offset, parent)) => {
+                                    // updating the parent node
+                                    if is_right {
+                                        parent.right = 0;
+                                    } else {
+                                        parent.left = 0;
+                                    }
+                                    self.storage.write(*offset, parent)?;
+                                }
+                                None => {
+                                    if node.left != 0 {
+                                        self.header.root_offset = node.left;
+                                    } else if node.right != 0 {
+                                        self.header.root_offset = node.right;
+                                    } else {
+                                        self.header.root_offset = 0
+                                    }
+                                } //
+                                  // TODO: deallocate item here
```

Comment:
will the TODO: deallocate item here will be a breaking change or just an enhancement in future?

Reply:
It will prevent storage leakage. After de-allocating a space in storage file, we can re-use it again. This part is not implemented yet.